File: simple_cpp_package/src/final_assessment.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/pose.hpp"
#include "geometry_msgs/msg/pose_array.hpp"
#include "rosi_msgs/srv/get_object_location.hpp"
#include <cmath>
#include <chrono>

using std::placeholders::_1;
using std::placeholders::_2;
using namespace std::chrono_literals;

class TurtleBotAruco : public rclcpp::Node
{
public:
    TurtleBotAruco() : Node("final_assessment")
    {
        distance_threshold_ = 1.0;
        max_linear_velocity_ = 0.1;
        max_angular_velocity_ = 1.0;
        too_close_distance_ = 0.15;
        alignment_angle_threshold_ = M_PI / 9.0; // 20 degrees

        cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
        twist_msg_ = geometry_msgs::msg::Twist();

        service_client_ = this->create_client<rosi_msgs::srv::GetObjectLocation>("/find_objects");

        while (!service_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(this->get_logger(), "Waiting for service...");
        }

        timer_ = this->create_wall_timer(100ms, std::bind(&TurtleBotAruco::send_request, this));
    }

private:
    void send_request()
    {
        auto request = std::make_shared<rosi_msgs::srv::GetObjectLocation::Request>();
        auto future = service_client_->async_send_request(request,
                    std::bind(&TurtleBotAruco::response_callback, this, std::placeholders::_1));
    }

    void response_callback(rclcpp::Client<rosi_msgs::srv::GetObjectLocation>::SharedFuture future)
    {
        try
        {
            auto result = future.get();
            auto pose = result->object_pose;

            RCLCPP_INFO(this->get_logger(), "Object Pose - x: %.2f, y: %.2f, z: %.2f",
                        pose.position.x, pose.position.y, pose.position.z);

            check_tag(result, pose);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed: %s", e.what());
        }
    }

    void check_tag(const rosi_msgs::srv::GetObjectLocation::Response::SharedPtr &result,
                   const geometry_msgs::msg::Pose &pose)
    {
        if (result->result)
        {
            RCLCPP_INFO(this->get_logger(), "robot can run");

            double current_distance = get_tag_distance(pose.position.x, pose.position.y);
            RCLCPP_INFO(this->get_logger(), "current_distance: %.2f", current_distance);

            double angle_to_tag = get_tag_angle(pose.position.x, pose.position.y);
            RCLCPP_INFO(this->get_logger(), "angle_to_tag (degree): %.2f", angle_to_tag * 180 / M_PI);

            if (std::abs(angle_to_tag) > alignment_angle_threshold_)
            {
                double angular_vel = angle_to_tag;
                if (angular_vel > max_angular_velocity_)
                    angular_vel = max_angular_velocity_;
                move_turtlebot(0.0, angular_vel);
                cmd_vel_pub_->publish(twist_msg_);
                return;
            }

            if (current_distance >= distance_threshold_)
            {
                RCLCPP_INFO(this->get_logger(), "\u2705 Reached target \u2192 stop");
                move_turtlebot(0.0, 0.0);
            }
            else if (current_distance <= too_close_distance_)
            {
                move_turtlebot(-max_linear_velocity_, 0.0);
            }
            else
            {
                RCLCPP_INFO(this->get_logger(), "\u27A1\uFE0F Moving toward ArUco");
                move_turtlebot(max_linear_velocity_, 0.0);
            }
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "robot can not run");
            move_turtlebot(0.0, 0.0);
        }

        cmd_vel_pub_->publish(twist_msg_);
        
    }

    double get_tag_distance(double x, double y)
    {
        return std::sqrt(x * x + y * y);
    }

    double get_tag_angle(double x, double y)
    {
        return std::atan2(y, x);
    }

    void move_turtlebot(double linear_velocity, double angular_velocity)
    {
        twist_msg_.linear.x = linear_velocity;
        twist_msg_.angular.z = angular_velocity;
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
    geometry_msgs::msg::Twist twist_msg_;
    rclcpp::Client<rosi_msgs::srv::GetObjectLocation>::SharedPtr service_client_;
    rclcpp::TimerBase::SharedPtr timer_;

    double distance_threshold_;
    double max_linear_velocity_;
    double max_angular_velocity_;
    double too_close_distance_;
    double alignment_angle_threshold_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleBotAruco>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: simple_python_package/simple_python_package/final_assessment.py
import rclpy
import math
from rclpy.node import Node
from geometry_msgs.msg import Twist, PoseArray, Pose
from perception_msgs.srv import GetObjectLocation
import numpy as np
import time

class TurtleBotAruco(Node):
    def __init__(self):
        super().__init__('final_assessment')
        # Recommended values, change if you want!
        self.distance_threshold = 1.0
        self.max_linear_velocity = 0.1
        self.max_angular_velocity = 1.0
        self.too_close_distance = 0.15
        self.alignment_angle_threshold =  math.radians(20) 
        # self.angular_gain = 2.0   


        self.cmd_vel_pub = self.create_publisher(Twist, 'cmd_vel', 10)
        self.twist_msg = Twist()

        self.service_client = self.create_client(GetObjectLocation, '/find_objects')
        while not self.service_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for service...')

        self.timer = self.create_timer(0.1, self.send_request) # thay đổi xuống còn 50ms để chạy nhanh hơn 


    def send_request(self):
        self.req = GetObjectLocation.Request()
        self.future_ = self.service_client.call_async(self.req)  
        self.future_.add_done_callback(self.response_callback)  

    def response_callback(self, future):
        try:
            result = future.result()
            # self.get_logger().info("Service response result: %d" % result.result)

            
            pose = result.object_pose
            self.get_logger().info(
                f"Object Pose - x: {pose.position.x:.2f}, y: {pose.position.y:.2f}, z: {pose.position.z:.2f}")

            # self.get_logger().info(
            #     f"Orierươcntation - x: {pose.orientation.x:.2f}, y: {pose.orientation.y:.2f}, z: {pose.orientation.z:.2f}, w: {pose.orientation.w:.2f}")

            self.check_tag(result, pose)
            

        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")

    def check_tag(self, result, pose):

        if result.result is True:
            print("robot can run")

             # Level 1
            # self.move_turtlebot(self.max_linear_velocity, 0.0)

            # Level 2
            current_distance = self.get_tag_distance(pose.position.x, pose.position.y)
            print("current_distance: ", current_distance)

            angle_to_tag = self.get_tag_angle(pose.position.x, pose.position.y)
            # print("angle_to_tag: ", angle_to_tag)
            print("angle_to_tag (degree):", math.degrees(angle_to_tag))

            if abs(angle_to_tag) > self.alignment_angle_threshold:
                angular_vel = angle_to_tag
                if angular_vel > self.max_angular_velocity:
                    angular_vel = self.max_angular_velocity
                self.move_turtlebot(0.0, angular_vel)
                self.cmd_vel_pub.publish(self.twist_msg)
                return 

            if current_distance >= self.distance_threshold:
                print("✅ Reached target → stop")
                self.move_turtlebot(0.0, 0.0)

            elif current_distance <= self.too_close_distance:
                self.move_turtlebot(-self.max_linear_velocity, 0.0)
            
            # elif angle_to_tag
            
            else:
                print("➡️ Moving toward ArUco")
                self.move_turtlebot(self.max_linear_velocity, 0.0)
             
        else:
            print("robot can not run")
            
            self.move_turtlebot(0.0, 0.0)


        self.cmd_vel_pub.publish(self.twist_msg)


    
    def get_tag_pose(self):
        pass
        #------------------------------------------------
        #                    TODO:
        #  Create a function to get the pose of the tag 
        #------------------------------------------------

    def move_turtlebot(self, linear_velocity, angular_velocity):
        pass
        #------------------------------------------------
        #                    TODO:
        #  Create a function to move the robot
        #------------------------------------------------
        self.twist_msg.linear.x = linear_velocity
        self.twist_msg.angular.z = angular_velocity

    def get_tag_distance(self, x, y):
        return math.sqrt((x ** 2) + (y ** 2))

    def get_tag_angle(self, x, y):
        return np.arctan(y/x)
    

    # def get_tag_angle(self, x, y):
    #     return np.arctan2(y, x)


def main(args=None):
    rclpy.init(args=args)
    turtlebot_move = TurtleBotAruco()
    rclpy.spin(turtlebot_move)
    # while(rclpy.ok()):
    #     turtlebot_move.check_tag()
    turtlebot_move.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
